modules/dns-helper/dns_local_server.go

```diff
@@ -0,0 +1,112 @@
+package dns_helper
+
+import (
+	"fmt"
+	"log"
+	"net"
+	"strings"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+var testDomain = "gruntwork.io"
+
+// DNSData type
+type DNSData map[DNSQuery][]DNSAnswer
+
+func setupTestDNSServers() (ns1, ns2 string, dnsData1, dnsData2 *DNSData) {
+	ns1, mux1 := runTestDNSServer("0")
+	ns2, mux2 := runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	dnsData1 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+	dnsData2 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+
+	mux1.HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) { stdDNSHandler(w, r, dnsData1, false) })
+	mux2.HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) { stdDNSHandler(w, r, dnsData2, true) })
+
+	return
+}
+
+func setupTestDNSServersRetry() (ns1, ns2 string, dnsData1, dnsData2, dnsDataRetry1, dnsDataRetry2 *DNSData) {
+	ns1, mux1 := runTestDNSServer("0")
+	ns2, mux2 := runTestDNSServer("0")
+
+	q := DNSQuery{"NS", testDomain}
+	dnsData1 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+	dnsData2 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+	dnsDataRetry1 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+	dnsDataRetry2 = &DNSData{q: DNSAnswers{{"NS", ns1}, {"NS", ns2}}}
+
+	mux1.HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) { retryDNSHandler(w, r, dnsData1, dnsDataRetry1, false) })
+	mux2.HandleFunc(testDomain+".", func(w dns.ResponseWriter, r *dns.Msg) { retryDNSHandler(w, r, dnsData2, dnsDataRetry2, true) })
+
+	return
+}
+
+func runTestDNSServer(port string) (string, *dns.ServeMux) {
+	listener, err := net.ListenPacket("udp", "127.0.0.1:"+port)
+
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	mux := dns.NewServeMux()
+	server := &dns.Server{PacketConn: listener, Net: "udp", Handler: mux}
+
+	go func() {
+		log.Fatal(server.ActivateAndServe())
```

Comment:
Don't use `log.Fatal` in test functions as it can kill the entire test run, which is problematic for terratest where it could skip the destroy call in other tests.

Also, how does this server shutdown during testing?

Reply:
Refactored the server setup to properly defer it's shutdown from the tests functions.